pkg/router/template/router.go

```diff
@@ -530,16 +559,168 @@ func (r *templateRouter) DeleteServiceUnit(id string) {
 	r.stateChanged = true
 }
 
+// addServiceAliasAssociation adds a reference to the backend in the ServiceUnit config.
+func (r *templateRouter) addServiceAliasAssociation(id, alias string) {
+	if serviceUnit, ok := r.findMatchingServiceUnit(id); ok {
+		glog.V(4).Infof("associated service unit %s -> service alias %s", id, alias)
+		serviceUnit.ServiceAliasAssociations[alias] = true
+	}
+}
+
+// removeServiceAliasAssociation removes the reference to the backend in the ServiceUnit config.
+func (r *templateRouter) removeServiceAliasAssociation(id, alias string) {
+	if serviceUnit, ok := r.findMatchingServiceUnit(id); ok {
+		glog.V(4).Infof("removed association for service unit %s -> service alias %s", id, alias)
+		delete(serviceUnit.ServiceAliasAssociations, alias)
+	}
+}
+
+// dynamicallyAddRoute attempts to dynamically add a route.
+// Note: The config should have been synced at least once initially and
+//       the caller needs to acquire a lock [and release it].
+func (r *templateRouter) dynamicallyAddRoute(backendKey string, route *routeapi.Route, backend *ServiceAliasConfig) bool {
+	if r.dynamicConfigManager == nil {
+		return false
+	}
+
+	glog.V(4).Infof("Dynamically adding route backend %s", backendKey)
+	wildcard := backend.IsWildcard
+	r.dynamicConfigManager.Register(backendKey, route, wildcard)
+
+	// If no initial sync was done, don't try to dynamically add the
+	// route as we will need a reload anyway.
+	if !r.synced {
```

Comment:
Why is this after registration?

Reply:
Because we don't add the routes if the initial sync has not been done. There's nothing in the haproxy config at that point in time (backends don't exist for blueprints as yet, etc).  Plus since there's going to be a reload and the config is going to change anyway, the configmanager is not really going to do anything here (other than registration). 
So, the registration is done always - irrespective of whether the router sync state. And that ensures that the mappings in the configmanager are correct.